src/operator/svm_output-inl.h

```diff
@@ -0,0 +1,231 @@
+                       const std::vector<TBlob> &aux_args) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(in_data.size(), 2) << "Expecting [data, label]";
+    CHECK_EQ(out_data.size(), 1) << "Expecting [output]";
+    CHECK_EQ(req.size(), 1) << "Expecting output.size() == req.size()";
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    Tensor<xpu, 2, DType> data = in_data[svm_enum::kData].FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> out = out_data[svm_enum::kOut].FlatTo2D<xpu, DType>(s);
+    out = DType(0.0f);
+    out += data;
+  }
+
+  virtual void Backward(const OpContext &ctx,
+                        const std::vector<TBlob> &out_grad,
+                        const std::vector<TBlob> &in_data,
+                        const std::vector<TBlob> &out_data,
+                        const std::vector<OpReqType> &req,
+                        const std::vector<TBlob> &in_grad,
+                        const std::vector<TBlob> &aux_args) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(in_data.size(), 2);
+    CHECK_EQ(out_grad.size(), 1);
+    CHECK_GE(in_grad.size(), 1);
+    CHECK_GE(req.size(), 1);
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    const TShape& label_shape = in_data[svm_enum::kLabel].shape_;
+
+    Tensor<xpu, 1, DType> label = in_data[svm_enum::kLabel].get_with_shape<xpu, 1, DType>(
+        Shape1(label_shape.ProdShape(0, label_shape.ndim())), s);
+    Tensor<xpu, 2, DType> out = out_data[svm_enum::kOut].FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> grad = in_grad[svm_enum::kData].FlatTo2D<xpu, DType>(s);
+
+    if (param_.use_linear) {
+      CHECK_EQ(grad.shape_, out.shape_) << "L1_SVM: shape mismatch";
+      for (index_t y = 0; y < grad.size(0); y++) {
+        const index_t k = static_cast<int>(label[y]);
+        for (index_t x = 0; x < grad.size(1); x++) {
+          if (x == k) {
+            grad[y][k] = -DType(DType(param_.margin) > out[y][k])
+            * DType(param_.regularization_coefficient);
+          } else {
+            grad[y][x] = DType(DType(param_.margin) > -out[y][x])
+            * DType(param_.regularization_coefficient);
+          }
+        }
+      }
+    } else {
+      CHECK_EQ(grad.shape_, out.shape_) << "L2_SVM: shape mismatch";
+      for (index_t y = 0; y < grad.size(0); y++) {
+        const index_t k = static_cast<int>(label[y]);
+        for (index_t x = 0; x < grad.size(1); x++) {
+          if (x == k) {
+            grad[y][k] = DType(param_.margin) > out[y][k] ?
+            2*(DType(param_.margin) - out[y][k]) : DType(0.0f);
+            grad[y][k] *= -DType(param_.regularization_coefficient);
+          } else {
+            grad[y][x] = DType(param_.margin) > -out[y][x] ?
+            (-2)*(DType(param_.margin) + out[y][x]) : DType(0.0f);
+            grad[y][x] *= -DType(param_.regularization_coefficient);
+          }
+        }
+      }
+    }
```

Comment:
This may only work for CPU. If we want to support GPU we need to move these code to the `.cc` file and add additional CUDA kernels to `.cu`